### ISSUE TITLE
Window focus doesn’t change in some situations

### ISSUE THREAD

**Issue by VuiMuich** [leftwm/leftwm#209]
When the pointer is above an area that is `never_Focus` the focus is not updated correctly.

In case of switching tags it stays on the focused window of the previous focused tag.

(In case of #197 it creates the new window in the background)

**Comment by AethanFoot** [comment]
I also get this weird behaviour as well, it is almost like the focus is sticky sometimes. If I remember correctly the border will change colour as if the focus has changed but the focus in the app doesn't update, however I'll try and do more testing on it to see if I can find a way to reproduce it.

**Comment by AethanFoot** [comment]
I have noticed when in the monocle layout that when cycling through focus only changes between the top two windows. Also when switching focus the mouse always goes to the far left of my left monitor, instead of the middle of the focused window (this then prevents switching on the main monitor as the cursor moves to the other screen).

**Comment by lex148** [comment]
I see this as well. I am testing a fix right now. should have a PR up later today.

**Comment by lex148** [comment]
This turned out not to be one problem but several little focus oddities. I have found about four today while testing. I have merged the fixes for them. If we continue to see focus bugs let me know and we will keep killing them.

**Comment by VuiMuich** [comment]
absolutely love the pace this project is moving right now :smiley:

**Comment by AethanFoot** [comment]
I am still having focus issues in the monocle layout. The biggest issue I am having is when focus changes when a workspace in this layout, the mouse cursor moves to the far left of my monitor at +0+0 (taking focus with it). Examples of when this happens is when cycling focus, and adding new windows. This may be a bug within the code mautamu added. If it is helpful I can record a clip to better show what happens.

**Comment by mautamu** [comment]
It most likely is. Probably need to add the function that centers the mouse on a window to the lines I put in.

**Comment by mautamu** [comment]
@AethanFoot try removing line 266:
```rust
let act = DisplayAction::MoveMouseOver(handle);```

**Comment by AethanFoot** [comment]
ok, so that helps when cycling through the windows (obviously as moving the mouse is no longer called). However there must be an issue with the handle when on monocle as the issue happens when a new window is spawned which shifts the focus and the mouse to the left of the monitor at +0+0 (but the new window is spawned hidden). 

Aside from the mouse, when cycling through the windows (in monocle) it will cycle going up properly, but will only switch between the original window and the one below when going down.

**Comment by AethanFoot** [comment]
To add to this, in more testing this also happens when moving the windows. So it is not an issue with focus but the moving mechanism. Plus when in the monocle layout should a new window be spawned at the top of the stack, or not take focus.

**Comment by VuiMuich** [comment]
except that it is lines 280 and 324 when you create a new branch from master (is this also called forking, btw?) and the collowing lines with
```manager.actions.push_back(act);```
need to go as well, it fixes FocusUp/Down for me.
Window-Creation is still bugged though..

**Comment by mautamu** [comment]
280/324 are implicated? That’s really odd, those are direct copies of the move handler.

**Comment by VuiMuich** [comment]
when I look at the upstream/master branch I find the line ```let act = DisplayAction::MoveMouseOver(handle);``` in line 280 and 324.

@AethanFoot: I think your EDIT2 is a good lead.
My preferred behaviour still would be that the monocle layout should strictly react on a change of focus: when I "move" a window I want to see a different window, when I move focus I want to see a different window, when I create a new window I most probably want to see this new window now etc. 

Maybe if the visibility in the monocle layout follows `DisplayEvent::FocusedWindow(WindowHandle, i32, i32)`, so it gets changed after the focus_handler called `DisplayAction::MoveMouseOver(WindowHandle)`

**Comment by mautamu** [comment]
Ah, thanks for the clarification! 

Is there a way to easily pass the focus history to the layouts?

**Comment by AethanFoot** [comment]
I think the best way of doing it would be to add a method to the manager that gives the top window, which ideally for at least the monocle layout would allow you to set focus to the top window which should be the visible one. Then we would want to make new windows spawn at the top of the stack (rather than at the bottom), then the focus and other stuff should be handled already.

**Comment by VuiMuich** [comment]
> Is there a way to easily pass the focus history to the layouts?

Not that I have found a way that works, but I have to admit my coding-abilities are limited.
As far as I understand either the layout's update function needs to be called with the manager as parameter or the focus_state is stored in a window (like visibility) and it has a method to return `pub fn has_focus (&mut self) -> bool`.
Another alternative (or maybe even an addition) might be to give the workspace a field that holds the focused window.

`DisplayEvent` and `DisplayAction` are just calls to xlib bindings if I got that right, so won't work either.

> I think the best way of doing it would be to add a method to the manager that gives the top window, which ideally for at least the monocle layout would allow you to set focus to the top window which should be the visible one. Then we would want to make new windows spawn at the top of the stack (rather than at the bottom, for the monocle layout), then the focus and other stuff should be handled already.

As far as I get it, the window creation already puts the new window on top of the window-stack, but it happens as a DisplayAction, so again the layout is unaware of this.

To fix the bug of the mouse pointer jumping (in my case it's not +0+0, btw. but y is somehwere in the middle of the screen, probably half the height of the window height of some other layout, just by guess), this might help in the command_handler:
```rust
if handle.visible() { 
    let act = DisplayAction::MoveMouseOver(handle);
    manager.actions.push_back(act);
    }
```
I can't test this right now, will look try it later or tomorrow.

**Comment by VuiMuich** [comment]
so I tried this:
```rust
let window = match manager.focused_window() {
                         Some(w) => w.clone(),
                         _ => {
                              return false;
                          }
                      };
                      if window.visible() {
                          let act = DisplayAction::MoveMouseOver(handle);
                          manager.actions.push_back(act);
                      }
```
but it keeps jumping to +0+height/2 (or is this +y+x?)

**Comment by AethanFoot** [comment]
I would of thought you could get the top window from to_reorder, and then move the mouse over that handle. As the code above would be still trying to focus over the previous window, as the new top hasn't got focus yet. (Or you could be cheep and use the move mouse to point method and move it to the centre of the workspace). I would try it but Im very unfamiliar with rust syntax.

**Comment by AethanFoot** [comment]
I have now also realised that cycling down would never work past the first two windows, as the top window swaps places with the second window, so when you then try and swap the new top it again just switches places with the previous top.

**Comment by AethanFoot** [comment]
I have found something that works (only if two windows are open, dont know why) which is basically getting the window with relative_find before we move it about. E.g.
```
                    let new_handle = match helpers::relative_find(&to_reorder, is_handle, -1) {
                        Some(h) => h.handle.clone(),
                        _ => {
                            return false;
                        }
                    };
```
Put before the reorder and mousing over the new_handle. Again however this seems to only work with two windows. Also the layout handling would need adding to the movement commands as well with this type of thing as the extra.

**Comment by lex148** [comment]
I am reopening this ticket. It seams there are still focus issues.

**Comment by VuiMuich** [comment]
Thanks @lex148. I feel quite sorry for putting this buggy PR on the main project obviously way too early.
On the other hand, I am very grateful for all the input and feedback, since I am learning a ton from it.

I keep trying to implement a more reliablae and hopefully elegant solution in [this branch](https://github.com/VuiMuich/leftwm/blob/layout_monocle2/src/layouts/monocle.rsl) of my fork.

**Comment by AethanFoot** [comment]
@VuiMuich Let me know on here if you want some extra testing on this issue. Im happy to put it through the ringer.

**Comment by VuiMuich** [comment]
I think this might have been fixed by PR #230 (maybe already in #213 or #224).

So I'm closing this.
